### PR TITLE
Partial parsing and loading window

### PR DESCRIPTION
**Pull request by Anastasia-Nikitina**

### CODE REVIEW

**Comment by Anastasia-Nikitina** [comment]
Запись экрана от 19 02 2023 19 41 48

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 1]
```diff
@@ -35,90 +31,88 @@ object ProjectParser {
...
             try {
-                val longName = imageName.toLong()
-                landmarks.putIfAbsent(
-                    longName,
-                    mutableListOf()
-                )
-                landmarks[longName]?.add(LandmarkFile(layers.last(), Path(it.path), extractUIDs(it.path)))
+                imageName.toLong()
+                outputs.add(OutputAfterPartialParsing(imageName, it.absolutePath, directoryName, kind))
+                if (!layers.contains(directoryName)) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt:78]: why is it done here? we need to add only once for dir

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 2]
```diff
@@ -154,14 +152,25 @@ object ProjectParser {
...
+            }
+        }
+        images.forEach { img ->
+            if (img.errors.isNotEmpty()) {
+                hasAnyErrors = true
             }
+            frames.add(FrameAfterPartialParsing(img, outputs.filter { img.name == it.name }))
         }
 
+        if (noSomeAlgorithmError(frames, layers)) hasAnyErrors = true
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: add {}

Comment (DmiitriyJarosh) [reply src/main/kotlin/solve/importer/ProjectParser.kt]: you use a lot of dirty funcs, it is a bad solution)

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 3]
```diff
@@ -154,14 +152,25 @@ object ProjectParser {
...
+
+        frames.sortWith(compareBy { it.image.name.toLong() })
+
         if (!isImagesExist) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt:158]: why do we check it so late?

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 4]
```diff
@@ -170,39 +179,47 @@ object ProjectParser {
             alertErrorFolderMessage()
         }
 
-        landmarks.forEach { (name, outputs) ->
-            frames.add(ImporterProjectFrame(ProjectFrame(name, Path(images[name]!!), outputs)))
+        return ProjectAfterPartialParsing(path, frames, hasAnyErrors)
+    }
+
+    fun fullParseDirectory(
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: maybe to other file?

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 5]
```diff
@@ -35,90 +31,88 @@ object ProjectParser {
...
-    private fun parseImages(folder: File, images: ObservableMap<Long, String>) {
+    private fun partialParseImages(folder: File, images: ObservableList<ImageAfterPartialParsing>) {
         val errorImages = mutableListOf<String>()
         folder.listFiles()?.forEach {
             val imageName = it.nameWithoutExtension
             try {
-                val longName = imageName.toLong()
-                images[longName] = it.absolutePath
+                imageName.toLong()
+                if (!possibleExtensions.contains(it.extension)) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: use if inside apply)) code duplication here

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 6]
```diff
@@ -35,90 +31,88 @@ object ProjectParser {
...
 
-    private fun parseImages(folder: File, images: ObservableMap<Long, String>) {
+    private fun partialParseImages(folder: File, images: ObservableList<ImageAfterPartialParsing>) {
         val errorImages = mutableListOf<String>()
         folder.listFiles()?.forEach {
             val imageName = it.nameWithoutExtension
             try {
-                val longName = imageName.toLong()
-                images[longName] = it.absolutePath
+                imageName.toLong()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: ??

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 7]
```diff
@@ -35,90 +31,88 @@ object ProjectParser {
...
         folder.listFiles()?.forEach {
             val imageName = it.nameWithoutExtension
             try {
-                val longName = imageName.toLong()
-                landmarks.putIfAbsent(
-                    longName,
-                    mutableListOf()
-                )
-                landmarks[longName]?.add(LandmarkFile(layers.last(), Path(it.path), extractUIDs(it.path)))
+                imageName.toLong()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: ??

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 8]
```diff
@@ -154,14 +152,25 @@ object ProjectParser {
...
             if (folder.name != IMAGE_DIRECTORY_NAME) {
-                parseOutputs(folder, layers, landmarks, errorFolders)
+                partialParseOutputs(folder, outputs, errorFolders, algorithms, layers)
             } else {
                 isImagesExist = true
-                parseImages(folder, images)
+                partialParseImages(folder, images)
+            }
+        }
+        images.forEach { img ->
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: map?

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 9]
```diff
@@ -132,16 +126,20 @@ object ProjectParser {
...
-    fun parseDirectory(path: String): ImporterProject? {
+    fun partialParseDirectory(path: String): ProjectAfterPartialParsing? {
         val directory = File(path)
-        val images = FXCollections.observableHashMap<Long, String>()
-        val landmarks = FXCollections.observableHashMap<Long, MutableList<LandmarkFile>>()
-        val frames = FXCollections.observableArrayList<ImporterProjectFrame>()
-        val layers = FXCollections.observableArrayList<ProjectLayer>()
+        val images = FXCollections.observableArrayList<ImageAfterPartialParsing>()
+        val outputs = FXCollections.observableArrayList<OutputAfterPartialParsing>()
+        val algorithms = mutableListOf<String>()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt:129]: where is it used? why do we have also layers here?

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 10]
```diff
@@ -170,39 +179,47 @@ object ProjectParser {
...
+            landmarks[longName] = mutableListOf()
+            it.outputs.forEach { output ->
+                val currentLayer = ProjectLayer(output.kind, output.algorithmName)
+                if (!layers.contains(currentLayer)) {
+                    layers.add((currentLayer))
+                }
+                landmarks[longName]?.add(LandmarkFile(currentLayer, Path(output.path), extractUIDs(output.path)))
             }
+            frames.add(
+                landmarks[longName]?.toList()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: some problems with new lines)

File: src/main/kotlin/solve/importer/controller/ImporterController.kt [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
-import solve.importer.model.ImporterProject
+import solve.importer.model.ProjectAfterPartialParsing
+import solve.project.model.Project
 import tornadofx.Controller
 import tornadofx.objectProperty
 
 class ImporterController : Controller() {
     val directoryPath = SimpleStringProperty()
-    val project = objectProperty<ImporterProject>()
+    val project = objectProperty<Project>()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/controller/ImporterController.kt]: we still need property?

File: src/main/kotlin/solve/importer/model/ProjectAfterPartialParsing.kt [hunk 12]
```diff
@@ -0,0 +1,3 @@
+package solve.importer.model
+
+data class ProjectAfterPartialParsing (val currentDirectory: String, val value: List<FrameAfterPartialParsing>, val hasAnyErrors: Boolean = false)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/model/ProjectAfterPartialParsing.kt]: what is value?

File: src/main/kotlin/solve/importer/view/ControlPanel.kt [hunk 13]
```diff
@@ -22,18 +25,20 @@ class ControlPanel : View() {
 
     private val importButtonModel = ButtonModel()
 
+    private val coroutineScope = CoroutineScope(Dispatchers.Main)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/solve/importer/view/ControlPanel.kt]: what if we close window? will it be cancelled? why do we separate it in variable if never use anywhere?

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: src/main/kotlin/solve/importer/ProjectParser.kt (hunk 4 above)
Comment (Anastasia-Nikitina) [reply src/main/kotlin/solve/importer/ProjectParser.kt]: why? this applies to parsing too

File: src/main/kotlin/solve/importer/ProjectParser.kt (hunk 9 above)
Comment (Anastasia-Nikitina) [reply src/main/kotlin/solve/importer/ProjectParser.kt:129]: layers are used in `noSomeAlgorithmError` to determine which algorithms a particular image does not have

File: src/main/kotlin/solve/importer/view/ControlPanel.kt [hunk 14]
```diff
@@ -72,24 +83,29 @@ class ControlPanel : View() {
         importer.close()
     }
 
-    private fun importAction(button: Button) {
-        val projectVal = controller.project.value
+    private fun showLoading() {
+        find<ImporterView>().replaceWith(
+            LoadingScreen::class
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/view/ControlPanel.kt]: Fails when I import project second time [gif](https://imgur.com/a/j6JBtVP)

File: src/main/kotlin/solve/importer/view/LoadingScreen.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+                    }
+                    contentDisplay = ContentDisplay.TOP
+                }
+            }
+            bottom {
+                button("Cancel") {
+                    BorderPane.setAlignment(this, Pos.BOTTOM_CENTER)
+                    maxWidth = 130.0
+                    action {
+                        close()
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/view/LoadingScreen.kt:41]: Closes main window somehow ¯\_(ツ)_/¯
fYyVWnblIY

File: src/main/kotlin/solve/importer/model/ProjectAfterPartialParsing.kt [hunk 16]
```diff
@@ -0,0 +1,3 @@
+package solve.importer.model
+
+data class ProjectAfterPartialParsing (val currentDirectory: String, val projectFrames: List<FrameAfterPartialParsing>, val hasAnyErrors: Boolean = false)
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/model/ProjectAfterPartialParsing.kt]: space

File: src/main/kotlin/solve/importer/model/OutputAfterPartialParsing.kt [hunk 17]
```diff
@@ -0,0 +1,5 @@
+package solve.importer.model
+
+import solve.project.model.LayerKind
+
+data class OutputAfterPartialParsing (val name: String = "", val path: String = "", val algorithmName: String = "", val kind: LayerKind, val errors: MutableList<String> = mutableListOf())
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/model/OutputAfterPartialParsing.kt]: Space, length, default values is not required

File: src/main/kotlin/solve/importer/model/ImageAfterPartialParsing.kt [hunk 18]
```diff
@@ -0,0 +1,3 @@
+package solve.importer.model
+
+data class ImageAfterPartialParsing (val name: String = "", val path: String = "", val errors: MutableList<String> = mutableListOf())
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/model/ImageAfterPartialParsing.kt]: space, length, default values is not required

File: src/main/kotlin/solve/importer/ProjectParser.kt (hunk 4 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/solve/importer/ProjectParser.kt]: Full parsing and partial parsing is very different

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 19]
```diff
@@ -35,90 +31,96 @@ object ProjectParser {
...
-            frame.errorMessage.add("The image has an incorrect extension")
+    private fun noSomeAlgorithmError(
+        frames: ObservableList<FrameAfterPartialParsing>,
+        layers: MutableList<String>
+    ): Boolean {
+        var hasAnyErrors = false
+        frames.forEach { frame ->
+            val diff = layers.minus(frame.outputs.map { output -> output.algorithmName }.toSet())
+            if (diff.isNotEmpty()) {
+                val missingAlgorithms = diff.toMutableList().toStringWithoutBrackets()
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: Mutability is not really used

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 20]
```diff
@@ -35,90 +31,96 @@ object ProjectParser {
...
-        fun addNoSomeAlgorithmErrorMessage() {
-            val missingAlgorithms = diffString.toMutableList().toStringWithoutBrackets()
-            val imageTimestamp = frame.importerFrame.timestamp
-            frame.errorMessage.add("There ${if (diff.count() == 1) "is" else "are"} no $missingAlgorithms for image $imageTimestamp")
-        }
-
-        if (diff.isNotEmpty()) {
-            frame.isImageErrored = true
-            addNoSomeAlgorithmErrorMessage()
+    private fun incorrectExtensionError(
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: This function checks for the error and fills list of ImageAfterPartialParsing at the same time. Decompose please

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 21]
```diff
@@ -154,55 +160,76 @@ object ProjectParser {
...
 
         if (!isImagesExist) {
+            val ownerWindow = importer.root.scene.window
             createAlertForError("The images folder is missing in the directory", ownerWindow)
             return null
         }
 
+        images.map { img ->
+            if (img.errors.isNotEmpty()) {
+                hasAnyErrors = true
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: hasAnyErrors = img.errors.isNotEmpty()

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 22]
```diff
@@ -154,55 +160,76 @@ object ProjectParser {
...
+            }
+            frames.add(FrameAfterPartialParsing(img, outputs.filter { img.name == it.name }))
+        }
+
+        if (noSomeAlgorithmError(frames, layers)) {
+            hasAnyErrors = true
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: hasAnyErrors = noSomeAlgorithmError(frames, layers)

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 23]
```diff
@@ -154,55 +160,76 @@ object ProjectParser {
...
+        }
+
+        frames.sortWith(compareBy { it.image.name.toLong() })
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt:172]: Why it is important?

File: src/main/kotlin/solve/importer/view/LoadingScreen.kt (hunk 15 above)
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/view/LoadingScreen.kt:41]: Will it cancel the task?

File: src/main/kotlin/solve/importer/view/ControlPanel.kt [hunk 24]
```diff
@@ -44,11 +47,19 @@ class ControlPanel : View() {
...
         isDisable = importButtonModel.disabled.value
         prefWidth = ButtonWidth
-        importButtonModel.disabled.onChange{ disableValue ->
+        importButtonModel.disabled.onChange { disableValue ->
             disableValue?.also { isDisable = it }
         }
+
         action {
-            importAction(this)
+            CoroutineScope(Dispatchers.Main).launch {
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/view/ControlPanel.kt]: Is it JavaFX application thread?

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 25]
```diff
@@ -154,55 +160,76 @@ object ProjectParser {
...
-        var hasAnyErrors = false
-        frames.forEach {
-            incorrectExtensionError(it)
-            noSomeAlgorithmError(it, layers)
-            if (it.isImageErrored) {
-                hasAnyErrors = true
+        project.projectFrames.forEach {
+            val longName = it.image.name.toLong()
+            landmarks[longName] = mutableListOf()
+            it.outputs.forEach { output ->
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: map?

**Review by Alexander-Ploskin — CHANGES_REQUESTED** [review]

File: src/main/kotlin/solve/importer/ProjectParser.kt (hunk 23 above)
Comment (Anastasia-Nikitina) [reply src/main/kotlin/solve/importer/ProjectParser.kt:172]: so that the images are displayed in the correct order

**Review by DmiitriyJarosh — APPROVED** [review]

File: src/main/kotlin/solve/importer/FullParserForImport.kt [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+import solve.project.model.Project
+import solve.project.model.ProjectFrame
+import solve.project.model.ProjectLayer
+import kotlin.io.path.Path
+
+object FullParserForImport {
+    fun fullParseDirectory(
+        project: ProjectAfterPartialParsing
+    ): Project {
+        val landmarks = FXCollections.observableHashMap<Long, MutableList<LandmarkFile>>()
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/FullParserForImport.kt]: Use standard not observable collections

File: src/main/kotlin/solve/importer/FullParserForImport.kt [hunk 27]
```diff
@@ -0,0 +1,39 @@
+package solve.importer
+
+import javafx.collections.FXCollections
+import solve.importer.model.ProjectAfterPartialParsing
+import solve.parsers.planes.ImagePlanesParser
+import solve.project.model.LandmarkFile
+import solve.project.model.Project
+import solve.project.model.ProjectFrame
+import solve.project.model.ProjectLayer
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/FullParserForImport.kt]: optimize imports

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 28]
```diff
@@ -132,16 +124,20 @@ object ProjectParser {
...
     }
 
-    fun parseDirectory(path: String): ImporterProject? {
+    fun partialParseDirectory(path: String): ProjectAfterPartialParsing? {
         val directory = File(path)
-        val images = FXCollections.observableHashMap<Long, String>()
-        val landmarks = FXCollections.observableHashMap<Long, MutableList<LandmarkFile>>()
-        val frames = FXCollections.observableArrayList<ImporterProjectFrame>()
-        val layers = FXCollections.observableArrayList<ProjectLayer>()
+        val images = FXCollections.observableArrayList<ImageAfterPartialParsing>()
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: use standard not observable collections

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 29]
```diff
@@ -132,16 +124,20 @@ object ProjectParser {
...
+        val outputs = FXCollections.observableArrayList<OutputAfterPartialParsing>()
+        val algorithms = mutableListOf<String>()
+        val frames = FXCollections.observableArrayList<FrameAfterPartialParsing>()
+        val layers = mutableListOf<String>()
+        var hasAnyErrors: Boolean
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: initialize variable

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 30]
```diff
@@ -154,55 +150,47 @@ object ProjectParser {
...
         }
 
-        var hasAnyErrors = false
-        frames.forEach {
-            incorrectExtensionError(it)
-            noSomeAlgorithmError(it, layers)
-            if (it.isImageErrored) {
-                hasAnyErrors = true
-            }
+        hasAnyErrors = noSomeAlgorithmError(frames, layers)
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: What if there hasAnyErrors was true, but `noSomeAlgorithmError(frames, layers)` is false?
hasAnyErrors = hasAnyErrors || noSomeAlgorithmError(frames, layers)

File: src/main/kotlin/solve/importer/ProjectParser.kt [hunk 31]
```diff
@@ -154,55 +150,47 @@ object ProjectParser {
...
+            val ownerWindow = importer.root.scene.window
             createAlertForError("The images folder is missing in the directory", ownerWindow)
             return null
         }
 
-        if (errorFolders.isNotEmpty()) {
-            alertErrorFolderMessage()
-        }
+        images.map { img ->
+            hasAnyErrors = img.errors.isNotEmpty()
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/importer/ProjectParser.kt]: The same as for noSomeAlgorithmError

File: src/main/kotlin/solve/utils/ImporterUtils.kt [hunk 32]
```diff
@@ -13,6 +13,11 @@ fun createAlertForError(content: String, owner: Window) {
     alert.show()
 }
 
+@JvmName("toStringForMutable")
 fun MutableList<String>.toStringWithoutBrackets(): String {
```
Comment (Alexander-Ploskin) [inline src/main/kotlin/solve/utils/ImporterUtils.kt]: Remove this function, MutableList is an inheritor for List

**Review by Alexander-Ploskin — CHANGES_REQUESTED** [review]

**Review by Alexander-Ploskin — APPROVED** [review]

**Pull request merged**